File: MasteringSwift/MasteringSwift/chapters/chapter15/advanced/AdvancedFunctional.swift
//
//  AdvancedFunctional.swift
//  MasteringSwift
//
//  Capítulo 15 — Programación funcional.
//  Nivel AVANZADO: StudentScoreSystem — map de puntuaciones, filter aprobados, reduce al promedio.
//

import Foundation

struct Student {
    let name: String
    let score: Int
}

func runAdvancedFunctionalExamples() {
    let students = [
        Student(name: "Alice", score: 85),
        Student(name: "Bob", score: 72),
        Student(name: "Charlie", score: 90),
        Student(name: "Diana", score: 58),
        Student(name: "Eve", score: 88)
    ]
    let passingThreshold = 60

    print("========================================")
    print("       STUDENT SCORE SYSTEM")
    print("========================================")
    print("Students: \(students)")
    print("Passing threshold: \(passingThreshold)")

    let scores = students.map(\.score)
    print("Scores (map): \(scores)")

    let passingScores = scores.filter { $0 >= passingThreshold }
    print("Passing scores (filter): \(passingScores)")

    let sum = passingScores.reduce(0, +)
    let average = passingScores.isEmpty ? 0.0 : Double(sum) / Double(passingScores.count)
    print("Average of passing: \(average)")

    let passingList = students.map(\.score).filter { $0 >= passingThreshold }
    let avgPassing = passingList.isEmpty ? 0.0 : Double(passingList.reduce(0, +)) / Double(passingList.count)
    print("Pipeline average (passing only): \(avgPassing)")
    print("========================================")
}

File: MasteringDart/lib/chapters/chapter15/advanced/advanced_functional.dart
// ignore_for_file: avoid_print
//
//  advanced_functional.dart
//  MasteringDart
//
//  Capítulo 15 — Programación funcional.
//  Nivel AVANZADO: StudentScoreSystem — map scores, filter aprobados, reduce a media.
//
//  Ejecutar con: dart run lib/chapters/chapter15/advanced/advanced_functional.dart

class Student {
  final String name;
  final int score;
  const Student(this.name, this.score);
  @override
  String toString() => 'Student($name, $score)';
}

void runAdvancedFunctionalExamples() {
  const students = [
    Student('Alice', 85),
    Student('Bob', 72),
    Student('Charlie', 90),
    Student('Diana', 58),
    Student('Eve', 88),
  ];
  const passingThreshold = 60;

  print('========================================');
  print('       STUDENT SCORE SYSTEM');
  print('========================================');
  print('Students: $students');
  print('Passing threshold: $passingThreshold');

  final scores = students.map((s) => s.score).toList();
  print('Scores (map): $scores');

  final passingScores = scores.where((s) => s >= passingThreshold).toList();
  print('Passing scores (filter): $passingScores');

  final sum = passingScores.isEmpty
      ? 0
      : passingScores.reduce((a, b) => a + b);
  final average =
      passingScores.isEmpty ? 0.0 : sum / passingScores.length;
  print('Average of passing: $average');

  final passingList = students
      .map((s) => s.score)
      .where((s) => s >= passingThreshold)
      .toList();
  final avgPassing = passingList.isEmpty
      ? 0.0
      : passingList.reduce((a, b) => a + b) / passingList.length;
  print('Pipeline average (passing only): $avgPassing');
  print('========================================');
}

void main() {
  runAdvancedFunctionalExamples();
}
